Reimplement operator definitions locally

Jackywathy opened the pull request:
Implements all the operator information in python in this library, rather than relying on an external webservice (r6unhash.tk).

Jackywathy committed:
update requirements for aiohttp

Danampasis committed:
fixed authentication, upgraded from v2 to v3

Danampasis committed:
Quick fix for rank data

Danampasis committed:
Delete auth.py

Jackywathy committed:
WIP: implemented some operator definitions in the definitions subpackage

- converters contain classes that convert the raw operator definition into a class
- info are data classes containing info about the operators and their loadouts
- loadouts.py and operators.py are raw definitions files (they're simply a dictionary)
- scripts to generate 
- altered players.py (only load_operator) to use the new definitions
- added placeholder for ARUNI in their unique statistic
- fixed some boolean operators - should be x not in rather than not x in which negates x first

Jackywathy committed:
update gitignore to remove virtual env files

Jackywathy committed:
Revert "updated example/old_asyncio_example to use await / async syntax"

This reverts commit a9f81d2456615ecedc8f6b8fbdabb9cfdd1d5e3e.

Jackywathy committed:
updated example/old_asyncio_example to use await / async syntax

Jackywathy committed:
added seasons and rank definitions

- the constants are now imported straight from the definitions package
- season.py / rank.py contain the raw data modelled as a python dictionary
- converters contain the basic logic required to convert the raw objects to models and then store it in a collection for easy retrieval later
-

Jackywathy committed:
removed legacy definitions calls from player.py

- note: still missing operator unique stat, this will have to be reimplemented in the future

Jackywathy committed:
removed legacy definition calls from rank.py

works for examples. but need more testing to ensure it works with all other seasons

Jackywathy committed:
removed all the definitions from auth

this is no longer needed since we can no longer use the old definition system that relied on unhash.tk

Jackywathy committed:
added unique abilities to each operator

Jackywathy committed:
re-implemented unique stats on load_operator / load_all_operator

- Operator now takes a third parameter containing the unique stats
- small tweaks and refactors to reuse code that splits and processes the data

Jackywathy committed:
bumped aiohttp version, removed dependency on dateutil

Jackywathy committed:
removed a few comments, added MIT to new files

Jackywathy committed:
also update requirements for dev

Jackywathy commented:
closes #67, closes #68 and closes Jackywathy/RainbowSixSiege-Python-API#2

Jackywathy commented:
@Danampasis @billy-yoyo to review?

billy-yoyo commented on examples/old_asyncio_example.py:
```diff
@@ -1,14 +1,14 @@
 import asyncio
```
Why'd you update this file? This is now identical to `simple_example.py`? This file isn't really relevant anymore, but it's a relic of when I was keen on keeping python 3.4 support for this library (which is why I avoid the async/await keywords throughout the codebase). There's probably no reason to do this anymore so we could update all the @asyncio.couroutine and yield from's to async and awaits if we wanted - but probably not worth the effort.

Jackywathy replied:
Oh I didn't realize that that was for the old asyncio support, I did a simple find and and replace operation. I would have liked to also keep the python 3.4 support but with python 3.10 [removes](https://docs.python.org/3/library/asyncio-task.html#asyncio.coroutine) this version so it might not be worth supporting it. Your call.

billy-yoyo replied:
Hmm should probably update to support python 3.10, I think supporting modern versions is more of a priority than supporting legacy versions.

But let's not do that under this PR, I'll make a new issue for it

Jackywathy replied:
okay I'll revert it

billy-yoyo commented on r6sapi/definitions/converters.py:
```diff
@@ -0,0 +1,221 @@
+
+class Seasons:
+    """
+
```
missing comment

billy-yoyo commented on r6sapi/definitions/converters.py:
```diff
@@ -0,0 +1,221 @@
+"""
```
The name of this file seems a bit misleading to me - reading the code it seems more like a collection of data class stores rather than converters? Maybe `stores.py` would better describe its function?

Jackywathy replied:
True! I'll switch it over.

Jackywathy replied:
Actually I plan to split out the converting functionality to a new file; and have the new `stores.py` simply store `Info` objects.

billy-yoyo commented on r6sapi/definitions/loadouts.py:
```diff
@@ -0,0 +1,1079 @@
+"""
```
I'm not opposed to having the definitions hard coded, but am conscious it results in the library being more sensitive to API / game changes (updates etc.), increasing the effort required to keep it up to date, whereas before I was trying to move in the opposite direction.

I think for now this solution is fine, but I think a solution where we try and fetch & cache definitions from online first, and if it fails we fallback on these might be better. It might not be worth it though.

Jackywathy replied:
Yep that is a major downside of having the definitions hard coded; However I cannot find any authoritative sources of this information (please ubi make your api public).  The definitions package is intentionally simple so if we end up doing that down the line, we can simply tweak the converters.py to use online -> hardcoded

billy-yoyo reviewed (changes requested):
Changes overall look really good - thanks for taking the time to clean some of the files up :)
Just a couple of minor things to change then I'm happy to merge

Jackywathy committed:
renamed converters.py to stores.py

the main use of the classes is to store and retrieve the models. They also convert  the raw dictionary to a model, so it may be worth moving that logic if required in the future

billy-yoyo commented:
Will do a bit of testing and then do a release later today
